### PR TITLE
Model Import

### PR DESCRIPTION
**Pull request by mr-phazer**
Edits to shaders. Removed some debugging code.

### CODE REVIEW

File: CommonControls/FileTypes/RigidModel/MaterialHeaders/WeightedMaterial.cs [hunk 1]
```diff
@@ -87,7 +87,7 @@ public uint ComputeSize()
                 stringParamSize += res.Length;
             }
 
-            var headerDataSize = (uint)(ByteHelper.GetSize<WeightedMaterialStruct>() +
+            var headerDataSize = (uint)(ByteHelper.GetSize<CommonMaterialStruct>() +
```
Comment (donkeyProgramming) [inline CommonControls/FileTypes/RigidModel/MaterialHeaders/WeightedMaterial.cs:90]: Why was this changed?

File: CommonControls/FileTypes/RigidModel/MaterialHeaders/WeightedMaterial.cs [hunk 2]
```diff
@@ -396,7 +427,10 @@ List<RmvVector4> LoadVec4Params(uint Vec4ParamCount, byte[] dataArray, ref int d
 
     }
 
-    struct WeightedMaterialStruct
+    /// <summary>
+    /// TODO: Renamed this struct from "WeightedMaterialStruct" as it the same for several RigidMaterials
+    /// </summary>
+    struct CommonMaterialStruct
```
Comment (donkeyProgramming) [inline CommonControls/FileTypes/RigidModel/MaterialHeaders/WeightedMaterial.cs:433]: Why is this needed?

File: CommonControls/FileTypes/RigidModel/MaterialHeaders/WeightedMaterial.cs [hunk 3]
```diff
@@ -205,14 +208,42 @@ public IMaterial Create(ModelMaterialEnum materialId, RmvVersionEnum rmvType, by
             return material;
         }
 
+        public IMaterial CreateEmpty(ModelMaterialEnum materialId, RmvVersionEnum rmvType, VertexFormat vertexFormat)
+        {
+            // TODO: DEBUGGIN: remove these 2 line where sure they are not needed for this
```
Comment (donkeyProgramming) [inline CommonControls/FileTypes/RigidModel/MaterialHeaders/WeightedMaterial.cs:213]: Dont checkin debug code

File: CommonControls/ModelImportExport/AssimpImporter.cs [hunk 4]
```diff
@@ -0,0 +1,372 @@
+﻿using Microsoft.Xna.Framework;
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Text;
+using CommonControls.FileTypes.RigidModel.LodHeader;
+using CommonControls.FileTypes.RigidModel.MaterialHeaders;
+using CommonControls.FileTypes.RigidModel.Transforms;
+using CommonControls.FileTypes.RigidModel.Types;
+using CommonControls.FileTypes.RigidModel;
```
Comment (donkeyProgramming) [inline CommonControls/ModelImportExport/AssimpImporter.cs:10]: Probably lots of headers that should not be here

File: CommonControls/ModelImportExport/AssimpImporter.cs [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+        {
+            _packFileService = packFileService;
+        }
+        public void ImportScene(string fileName)
+        {
+            ImportAssimpScene(fileName);
+        }
+        private string GetSkeletonString()
+        {
+            string tempSkeletonString = new string("");
```
Comment (donkeyProgramming) [inline CommonControls/ModelImportExport/AssimpImporter.cs:48]: There is no need for new string(). Do "" or string.empty

File: CommonControls/ModelImportExport/AssimpImporter.cs [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+
+            var parent = _assScene.RootNode;
+
+            SearchNodesRecursiveLocal(parent, ref tempSkeletonString);
+
+            return tempSkeletonString;
+
+            void SearchNodesRecursiveLocal(Node parent, ref string skeletonString)
```
Comment (donkeyProgramming) [inline CommonControls/ModelImportExport/AssimpImporter.cs:56]: Functions should not be nested

File: CommonControls/ModelImportExport/AssimpImporter.cs [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+            {
+                foreach (var node in parent.Children)
+                {
+                    if (node.Name.Contains("skeleton"))
```
Comment (donkeyProgramming) [inline CommonControls/ModelImportExport/AssimpImporter.cs:60]: Personal preference, but consider not using {} for one liners. It makes the reader scroll more without adding any real value with how modern IDEs work

File: CommonControls/ModelImportExport/AssimpImporter.cs [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+using static System.Windows.Forms.VisualStyles.VisualStyleElement;
+using CommonControls.FileTypes.PackFiles.Models;
+using CommonControls.Services;
+using SharpDX.Direct3D9;
+using Serilog;
+using CommonControls.Common;
+using SharpDX.MediaFoundation;
+
+namespace CommonControls.ModelImportExport
+{
```
Comment (donkeyProgramming) [inline CommonControls/ModelImportExport/AssimpImporter.cs:29]: This file in general has inconsistent whitespace use

File: CommonControls/ModelImportExport/AssimpImporter.cs [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+                    if (skeletonString.Length > 0)
+                    {
+                        return;
+                    }
+
+                    SearchNodesRecursiveLocal(node, ref skeletonString);
+                }
+            }
+        }
+        private void LoadSkeletonFile()
```
Comment (donkeyProgramming) [inline CommonControls/ModelImportExport/AssimpImporter.cs:74]: Have a look at SkeletonAnimationLookUpHelper class, it might make this easier

File: CommonControls/ModelImportExport/AssimpImporter.cs [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+                {
+                    _fileType = Encoding.ASCII.GetBytes("RMV2"),
+                    SkeletonName = _skeletonFile != null ? _skeletonFile.Header.SkeletonName : "",
+                    Version = RmvVersionEnum.RMV2_V6,
+                    LodCount = (uint)lodCount,
+                },
+                LodHeaders = new RmvLodHeader[lodCount],
+            };
+
+            // TODO: remove these if init above works right
```
Comment (donkeyProgramming) [inline CommonControls/ModelImportExport/AssimpImporter.cs:119]: TODO should not be part of the PR

File: CommonControls/PackFileBrowser/ContextMenuHandler.cs [hunk 11]
```diff
@@ -127,11 +136,61 @@ void OnAddFilesCommand()
...
+
                     var packFile = new PackFile(fileName, new MemorySource(File.ReadAllBytes(file)));
+
                     _packFileService.AddFileToPack(_selectedNode.FileOwner, parentPath, packFile);
                 }
             }
         }
+        void OnAddFilesWithConversionCommand()
+        {
+            var parentPath = _selectedNode.GetFullPath(); // get pack path, at mouse pointer
```
Comment (donkeyProgramming) [inline CommonControls/PackFileBrowser/ContextMenuHandler.cs:148]: This logic does not belong here. It should be in a different class. 
This layer should as much as possible try to only route information.

File: CommonControls/PackFileBrowser/ContextMenuHandler.cs [hunk 12]
```diff
@@ -405,6 +465,8 @@ ContextMenuItem GetItem(ContextItems type)
                     return new ContextMenuItem() { Name = "Create" };
                 case ContextItems.AddFiles:
                     return new ContextMenuItem() { Name = "Add file", Command = AddFilesCommand }; ;
+                case ContextItems.AddFilesWithConversion:
```
Comment (donkeyProgramming) [inline CommonControls/PackFileBrowser/ContextMenuHandler.cs:468]: Good job expanding on this system. Its pretty messy :)

File: View3D/Content/Shaders/Phazer/MetalRoughness_main.fx [hunk 13]
```diff
@@ -180,7 +180,14 @@ struct PixelInputType
     float3 normal : NORMAL0;
     float3 tangent : TANGENT;
     float3 binormal : BINORMAL;
-
+    
+    // TODO: REMOVE THID DEBUGGIN CODE
```
Comment (donkeyProgramming) [inline View3D/Content/Shaders/Phazer/MetalRoughness_main.fx:184]: TODO.

Comment (mr-phazer) [reply View3D/Content/Shaders/Phazer/MetalRoughness_main.fx:184]: I am stealing fiddling with the shaders, for debugging the imported model, so can I do this laer?

File: CommonControls/FileTypes/Animation/AnimationFile.cs [hunk 14]
```diff
@@ -169,6 +171,20 @@ static AnimationHeader GetAnimationHeader(ByteChunk chunk)
             return header;
         }
 
+        public int GetIdFromBoneName(string name)
+        {
+            var result = Bones.Where(x => x.Name.Equals(name, StringComparison.OrdinalIgnoreCase));
```
Comment (donkeyProgramming) [inline CommonControls/FileTypes/Animation/AnimationFile.cs:176]: This should also have ignore case

File: CommonControls/FileTypes/Animation/AnimationFile.cs [hunk 15]
```diff
@@ -169,6 +171,20 @@ static AnimationHeader GetAnimationHeader(ByteChunk chunk)
...
+
+            var boneInfo = result.FirstOrDefault();
+
+            if (boneInfo == null)
+            {
+                return -1;
+            }
+
+            return boneInfo.Id;
+
```
Comment (donkeyProgramming) [inline CommonControls/FileTypes/Animation/AnimationFile.cs:186]: new line after return is just wast of space

File: CommonControls/FileTypes/RigidModel/MaterialHeaders/WeightedMaterial.cs [hunk 16]
```diff
@@ -396,7 +424,10 @@ List<RmvVector4> LoadVec4Params(uint Vec4ParamCount, byte[] dataArray, ref int d
 
     }
 
-    struct WeightedMaterialStruct
+    /// <summary>
+    /// TODO: Renamed this struct from "WeightedMaterialStruct" as it the same for several RigidMaterials
```
Comment (donkeyProgramming) [inline CommonControls/FileTypes/RigidModel/MaterialHeaders/WeightedMaterial.cs:428]: Todo, should not be part of the checking

File: CommonControls/FileTypes/RigidModel/MaterialHeaders/WeightedMaterial.cs (hunk 2 above)
Comment (donkeyProgramming) [reply CommonControls/FileTypes/RigidModel/MaterialHeaders/WeightedMaterial.cs:430]: While its true that this is shared, the code is not written in such a way that it is. Hence it should keep its old name. 
If the code is changed in the future (and different PR) the struct can be renamed

File: CommonControls/FileTypes/RigidModel/RmvFile.cs [hunk 17]
```diff
@@ -65,5 +67,8 @@ void UpdateModelHeader(RmvModel model, RmvVersionEnum rmvVersion)
 
             model.CommonHeader = header;
         }
+
```
Comment (donkeyProgramming) [inline CommonControls/FileTypes/RigidModel/RmvFile.cs:70]: Lots of whitespace

File: CommonControls/FileTypes/RigidModel/RmvFileHeader.cs [hunk 18]
```diff
@@ -29,7 +29,8 @@ public string SkeletonName
                 if (result == false)
                     throw new Exception();
                 return Util.SanatizeFixedString(value);
-            }
+
```
Comment (donkeyProgramming) [inline CommonControls/FileTypes/RigidModel/RmvFileHeader.cs:32]: Whitespace

File: CommonControls/ModelImportExport/AssimpExporter.cs [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Serilog;
+using Accessibility;
+using CommonControls.FileTypes.RigidModel;
+using CommonControls.FileTypes.RigidModel.Vertex;
+using Assimp.Unmanaged;
+using Assimp;
+using System.Net.WebSockets;
```
Comment (donkeyProgramming) [inline CommonControls/ModelImportExport/AssimpExporter.cs:12]: There are some very strange headers here 
using System.Net.WebSockets;
using Accessibility;

Are they needed?

File: CommonControls/ModelImportExport/AssimpExporter.cs [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+
+namespace CommonControls.ModelImportExport
+{
+
```
Comment (donkeyProgramming) [inline CommonControls/ModelImportExport/AssimpExporter.cs:16]: Formatting of this file is all over the place

File: CommonControls/ModelImportExport/AssimpImporter.cs [hunk 21]
```diff
@@ -0,0 +1,371 @@
...
+using System.Windows.Forms;
+using System.IO;
+using Filetypes.ByteParsing;
+using static System.Windows.Forms.VisualStyles.VisualStyleElement;
+using CommonControls.FileTypes.PackFiles.Models;
+using CommonControls.Services;
+using SharpDX.Direct3D9;
+using Serilog;
+using CommonControls.Common;
+using SharpDX.MediaFoundation;
```
Comment (donkeyProgramming) [inline CommonControls/ModelImportExport/AssimpImporter.cs:26]: Lots of headers which probably should not be here

File: View3D/Content/Shaders/Phazer/MetalRoughness_main.fx (hunk 13 above)
Comment (donkeyProgramming) [reply View3D/Content/Shaders/Phazer/MetalRoughness_main.fx:184]: No, a PR should reflect a finished unit of work. Hence no TODO or DEBUG stuff

**Review by donkeyProgramming — CHANGES_REQUESTED** [review]
